### PR TITLE
[colorLib] Sort paints by decreasing number of layers when encoding

### PR DESCRIPTION
**Pull request by behdad**
Something like this. Untested. Trying to test now.

Fixes https://github.com/fonttools/fonttools/issues/3250

### CODE REVIEW

**Comment by behdad** [comment]
```
-rw-r--r--. 1 behdad behdad  4568244 Aug  9 13:10 NotoColorEmoji.ttf.after
-rw-r--r--. 1 behdad behdad  4590708 Aug  9 12:44 NotoColorEmoji.ttf.before
```

**Comment by behdad** [comment]
Saves some 20kb.

**Comment by behdad** [comment]
Wait. My measurement might be flawed.

**Comment by behdad** [comment]
In my testing with NotoColorEmoji-Regular, it saved a meager 824 bytes. Still, it's the right thing to do.

File: Lib/fontTools/colorLib/builder.py [hunk 1]
```diff
@@ -639,7 +639,11 @@ def buildColrV1(
     errors = {}
     baseGlyphs = []
     layerBuilder = LayerListBuilder(allowLayerReuse=allowLayerReuse)
-    for baseGlyph, paint in colorGlyphItems:
+    for baseGlyph, paint in sorted(
+        colorGlyphItems,
+        key=lambda item: len(getattr(item[0], "Layers", [])),
```
Comment (anthrotype) [inline Lib/fontTools/colorLib/builder.py:644]: instead of reverse=True you could add a `-` minus sign to the `len`, no?

**Comment by anthrotype** [comment]
The BaseGlyphList must be sorted by glyphID to be able to do the binary search, remember?

https://learn.microsoft.com/en-us/typography/opentype/spec/colr#baseglyph-and-layer-records

> The BaseGlyph records must be sorted in increasing glyphID order. It is assumed that a binary search can be used to find a matching BaseGlyph record for a specific glyphID.

If you really want to build the LayerList in that decreasing length order, you can do that but then you need to make sure that the baseGlyphs are sorted by glyphOrder at the end

**Comment by behdad** [comment]
You are right. I'll fix up.

**Comment by behdad** [comment]
@anthrotype PTAL.

**Review by anthrotype — APPROVED** [review]

**Comment by anthrotype** [comment]
would also be nice to have a test that exercise this if not much trouble

**Comment by behdad** [comment]
> would also be nice to have a test that exercise this if not much trouble

Good thing I did. Implementation was wrong. Building noto color emoji now to get new numbers.

**Comment by behdad** [comment]
@anthrotype PTAL again. Sorry about that earlier.

**Comment by behdad** [comment]
Humm. I don't see this making any difference in NotoColorEmoji. Investigating.

**Comment by behdad** [comment]
> Humm. I don't see this making any difference in NotoColorEmoji. Investigating.

I fixed the culprit. However, now the font *grows* by a few hundred bytes. Maybe status quo is good enough.

**Pull request closed without merging**